File: backend/src/main/java/com/fitnexus/backend/service/RazorpayService.java
package com.fitnexus.backend.service;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
public class RazorpayService {

    @Value("${razorpay.key-id:}")
    private String keyId;

    @Value("${razorpay.key-secret:}")
    private String keySecret;

    private final HttpClient httpClient = HttpClient.newHttpClient();
    private static final Pattern ORDER_ID = Pattern.compile("\"id\"\\s*:\\s*\"(order_[^\"]+)\"");

    public boolean isConfigured() {
        return keyId != null && !keyId.isBlank() && keySecret != null && !keySecret.isBlank();
    }

    public String getKeyId() {
        return keyId;
    }

    /** Creates a Razorpay order and returns its id. Amount is in the smallest currency unit (paise). */
    public String createOrder(long amountPaise, String currency, String receipt) {
        if (!isConfigured()) {
            throw new IllegalStateException("Razorpay is not configured on the server");
        }
        try {
            String auth = Base64.getEncoder().encodeToString((keyId + ":" + keySecret).getBytes(StandardCharsets.UTF_8));
            String body = "{\"amount\":" + amountPaise + ",\"currency\":\"" + currency
                    + "\",\"receipt\":\"" + receipt + "\",\"payment_capture\":1}";
            HttpRequest request = HttpRequest.newBuilder(URI.create("https://api.razorpay.com/v1/orders"))
                    .header("Authorization", "Basic " + auth)
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 300) {
                throw new RuntimeException("Razorpay order creation failed (" + response.statusCode() + "): " + response.body());
            }
            Matcher m = ORDER_ID.matcher(response.body());
            if (!m.find()) {
                throw new RuntimeException("Razorpay response did not contain an order id");
            }
            return m.group(1);
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("Could not reach Razorpay: " + e.getMessage(), e);
        }
    }

    /** Verifies the payment signature: HMAC_SHA256(orderId + "|" + paymentId, keySecret) == signature. */
    public boolean verifySignature(String orderId, String paymentId, String signature) {
        if (orderId == null || paymentId == null || signature == null) return false;
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(keySecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] hash = mac.doFinal((orderId + "|" + paymentId).getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString().equals(signature);
        } catch (Exception e) {
            return false;
        }
    }
}

File: backend-fastapi/app/services/razorpay_service.py
import base64
import hashlib
import hmac
import re

import httpx

from app.core.config import settings

_ORDER_ID_RE = re.compile(r'"id"\s*:\s*"(order_[^"]+)"')


def is_configured() -> bool:
    return bool(settings.razorpay_key_id) and bool(settings.razorpay_key_secret)


def create_order(amount_paise: int, currency: str, receipt: str) -> str:
    if not is_configured():
        raise RuntimeError("Razorpay is not configured on the server")
    auth = base64.b64encode(f"{settings.razorpay_key_id}:{settings.razorpay_key_secret}".encode()).decode()
    body = {"amount": amount_paise, "currency": currency, "receipt": receipt, "payment_capture": 1}
    resp = httpx.post(
        "https://api.razorpay.com/v1/orders",
        headers={"Authorization": f"Basic {auth}", "Content-Type": "application/json"},
        json=body,
        timeout=15,
    )
    match = _ORDER_ID_RE.search(resp.text)
    if not match:
        raise RuntimeError(f"Razorpay order creation failed: {resp.text}")
    return match.group(1)


def verify_signature(order_id: str, payment_id: str, signature: str) -> bool:
    if not is_configured() or not order_id or not payment_id or not signature:
        return False
    payload = f"{order_id}|{payment_id}".encode()
    expected = hmac.new(settings.razorpay_key_secret.encode(), payload, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)
